Add Flask 2.x compatibility + fix line number detection under python <= 3.8 and python 3.11

ValentinFrancois opened the pull request:
## Issue 1
### Cause
When installing `Flask` 2.0 with python > 3.6, the chosen version of its `Jinja2` dependency is >= `3.1.0`:

```
$ docker run -it --rm python:3.7 bash
root@44ff1cecc27d:/# pip3 install Flask==2.0
Collecting Flask==2.0
  Downloading Flask-2.0.0-py3-none-any.whl (93 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 93.2/93.2 KB 2.5 MB/s eta 0:00:00
Collecting Jinja2>=3.0
  Downloading Jinja2-3.1.2-py3-none-any.whl (133 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 133.1/133.1 KB 7.5 MB/s eta 0:00:00
```

With python 3.6, the installed `Jinja2 ` dependency seems to be < `3.1.0 ` which should not be problematic as I'll explain next.
```
$ docker run -it --rm python:3.6 bash
root@7a7dccf967f2:/# pip3 install Flask==2.0
Collecting Flask==2.0
  Downloading Flask-2.0.0-py3-none-any.whl (93 kB)
     |████████████████████████████████| 93 kB 1.6 MB/s 
Collecting Jinja2>=3.0
  Downloading Jinja2-3.0.3-py3-none-any.whl (133 kB)
     |████████████████████████████████| 133 kB 6.4 MB/s 
```

### Consequences
`import flask_selfdoc` will then raise an `ImportError` because it tries to import the follow function and class that were deprecated in `Jinja2` 3.0 and removed in `Jinja2` 3.1:
- [jinja2.evalcontextfilter](https://jinja.palletsprojects.com/en/3.0.x/api/#jinja2.evalcontextfilter) -> moved to `jinja2.pass_eval_context`
- [jinja2.Markup](https://tedboy.github.io/jinja2/generated/generated/jinja2.Markup.html) -> moved to `jinja2.utils.markupsafe.Markup`

## Issue 2
### Cause
The way that the line number of a decorator is detected changed across python versions:
- python <= 3.8: `stack()[1].lineno` points to the line above the decorated function (= to the closest decorator, not necessarily the one that did the call to `stack()`)
- python 3.9 and 3.10:  `stack()[1].lineno` points to the line of the decorated function
- python 3.11:  `stack()[1].lineno` points to the exact line of the decorator that did the call to `stack()`

#### Example:

```
    1   |def call_stack_and_get_lineno():
    2   |
    3   |    def decorator(func):
    4   |        calling_frame = stack()[1]
    5   |        print(calling_frame.lineno)
    6   |        return func
    7   |
    8   |    return decorator
    9   |
    10  |
    11  |@decorator1
    12  |@call_stack_and_get_lineno
    13  |@decorator2
    14  |def func():
    15  |    pass
```

- python <= 3.8: will print line 13
- python 3.9 and 3.10: will print line 14 (desired behaviour)
- python 3.11: will print line 12

### Consequences
- unit tests don't pass for python 3.11
- unit tests pass for python <= 3.8 but only because we adjust them (add a `-1` offset to the detected line number)


## What this PR does
### Issue 1
- Avoid `Jinja2`-related `ImportError` by trying the old and the new import for each Jinja function/class that is now deprecated
- Remove the check `Flask < 2.0` from the package dependencies
- Adapt the ranges of tested python and `Flask` versions in the unit tests (WIP)
  - `minimal-test.yml`: remove python 3.7 (security support ends in 4 months), add python 3.10
  - `test.yml`: add `Flask` 2.1 (not compatible with python 3.6), add python 3.10 and 3.11 (not compatible with `Flask` 1.0)
 - Adapt the import of the Flask app context to version 2.3 where  `_app_ctx_stack` is removed
 - Install target Flask version for tests using `pip3` (easier than with `poetry`) and manually override its dependency versions when they're too recent (`itsdangerous`, `Jinja2` and `MarkupSafe` for Flask 1.X)
 - List all installed packages before running tests
### Issue 2
- Fix line number detection by reading the source file and finding the next line starting with `def `

jwg4 commented:
Thanks for this.

I am very in favour in principle; I want to make sure that we do this correctly and that we don't needlessly limit compatibility with older versions.

ValentinFrancois commented:
@jwg4 cool, I clarified the PR description a bit.

I think I spotted an issue in `minimal-test.yml` and `test.yml` that was overwriting the target `Flask` version from the matrix. 

I also added python 3.10 to see which is the oldest supported version of `Flask` (I already excluded `1.0` based on your `README.md`).

jwg4 commented:
1. looks good in principle
2. thank you for sorting out the jinja2 mess which I was dreading!
3. removing Python 3.7 from the quick test matrix seems fine.
4. adding flask and python versions to the main test matrix and specifying not to run certain toxic combinations also seems like the right decision.
5. I wonder if we should add 'latest' to the Python version specs, any thoughts?
6. as you can see above, some of the combinations fail - it seems that poetry will not allow you to specify a version which has a more restrictive Python version than what you have set for your own project, even if your Python version matches both specs. This seems right in principle, but I wonder if there's a way of overriding it for these purposes?
7. the fact that 6 has only now come up suggests that you are right about the flask versions not actually being correct before your fix.
I think this should be merged as soon as we can find a solution or a workaround for 6.

jwg4 commented:
thank you so much for this, I have left several comments. let me know if any that don't make sense or you don't think belong to this change.

ValentinFrancois commented on .github/workflows/minimal-test.yml:
```diff
@@ -7,22 +7,40 @@ jobs:
     steps:
-      - uses: actions/checkout@v2
-      - uses: actions/setup-python@v2
+      - uses: actions/checkout@v3
```
GitHub was encouraging us to update to the most recent action versions, seems harmless

ValentinFrancois commented on .github/workflows/minimal-test.yml:
```diff
@@ -7,22 +7,40 @@ jobs:
+      - name: Force Python version (linux/macOS)
+        if: matrix.os != 'windows-latest'
+        # sed command for macOS: https://stackoverflow.com/a/44864004
+        run: sed -i.bak 's/python = "^3.6"/python = "~${{ matrix.python-version }}"/' pyproject.toml
```
I couldn't find a better way than sed-replacing the python version in `pyproject.toml` to the fixed, desired 3.X version.

ValentinFrancois commented on .github/workflows/minimal-test.yml:
```diff
@@ -7,22 +7,40 @@ jobs:
         run: poetry install
+
+      - name: List installed package versions for manual inspection
+        run: poetry --version && poetry show
```
just to control what packages we exactly installed in the end

ValentinFrancois commented on .github/workflows/test.yml:
```diff
@@ -7,22 +7,140 @@ jobs:
+        run: poetry add Flask==${{ matrix.flask-version }}
+
+      - name: Overwrite Flask dependencies for legacy install
+        if: matrix.flask-version < '2.0' && matrix.flask-version != 'latest'
```
some packages that are now too recent when installed via the Flask 1.X dependencies

ValentinFrancois commented on .github/workflows/test.yml:
```diff
@@ -7,22 +7,140 @@ jobs:
       - name: Run doctests
         run: poetry run doctest
+
+  check_pip_install:
```
add a different test strategy: 
- given a specific Python runtime version
- given a specific Flask version installed
- try installing `flask-selfdoc` via `pip` and check that we can use it without `ImportError`
=> this basically tests the dependency configuration in `pyproject.toml`

ValentinFrancois commented on .github/workflows/test.yml:
```diff
@@ -7,22 +7,140 @@ jobs:
+
+      - name: Install current flask-selfdoc from GitHub
+        run: |
+          git_url=$(echo "${{ github.repositoryUrl }}@${{ github.head_ref }}" | sed "s/git:/git+https:/")
```
nice way to install from the latest commit of the current PR branch

ValentinFrancois commented on .github/workflows/test.yml:
```diff
@@ -7,22 +7,140 @@ jobs:
+        run: pip3 install werkzeug==2.0.3
+
+      - name: List installed package versions for manual inspection
+        run: python3 --version && pip3 list
```
control what package versions we exactly installed

ValentinFrancois commented on .github/workflows/test.yml:
```diff
@@ -7,22 +7,140 @@ jobs:
+      - name: List installed package versions for manual inspection
+        run: python3 --version && pip3 list
+
+      - name: Check that Flask version did not change
```
control that `pip`'s dependency solver didn't need to update the already installed version of Flask

ValentinFrancois commented on pyproject.toml:
```diff
@@ -7,7 +7,11 @@ license = "MIT"
 [tool.poetry.dependencies]
 python = "^3.6"
-Flask = "^1.0"
+Flask = [
```
express Flask dependency conditionally (depending on the installed version of Python)

ValentinFrancois replied:
I think this solves point 6 that you mentioned here: https://github.com/jwg4/flask-selfdoc/pull/72#issuecomment-1435263813

However it is still not possible to rely on the own dependencies config of Flask 1.X because it accepts too recent versions of `itsdangerous`, `Jinja2` and `MarkupSafe` that cause `ImportError` etc.
So these have to be manually overwritten with older, compatible versions like I did in `test.yml`

ValentinFrancois commented:
@jwg4 sorry I didn't see your previous comments while I was still struggling to solve the issues with poetry & dependency handling.
I think I've come to a good solution, I left a few comments explaining my recent changes.

The only thing I'm not sure about is why the output of `flask_selfdoc` gets different line numbers depending on the installed python/Flask versions. For now I added some pre-processing function to remove the line numbers from the compared outputs in `test_check_example.py` but maybe you'll want to investigate this deeper.

ValentinFrancois commented:
I tried adding tests for latest python as you suggested (`3.x`, currently evaluating to `3.11.2`) but under this version, the unit tests for line numbers don't pass:

![image](https://user-images.githubusercontent.com/24223132/220121370-e458591f-00c6-464a-aa3d-4fe98d25d34f.png)


So I think there's something defninitely weird with `inspect` and how it calculates line numbers of caller frames.
Even with a specific python version, the tests can fail depending on the OS because of different line numbers:

![image](https://user-images.githubusercontent.com/24223132/220121574-7ca43aa3-159b-4263-b811-63167b9c11c7.png)

I'm removing the code I had added to ignore these line number differences, because it seems to be more important than I first thought.
By the way, I rewrote `test.yml` to install the target Flask version with `pip3`, which in the end was easier to use (especially under windows).

jwg4 commented:
My understanding is that the code which retrieves the line number where a function is defined has changed to take decorators into account in a different way. I believe this is a change in a python built-in which flask uses and does not have control over. I don't know how OS type affects this.

ValentinFrancois commented:
Interesting, now python 3.11 seems to have changed again the way line number is calculated.

In this unit test, the frame now points to this line:
![image](https://user-images.githubusercontent.com/24223132/220319799-63dca5d2-0e92-4236-be9d-83925cb551e7.png)

I suppose this is related to this: https://docs.python.org/3/whatsnew/3.11.html#whatsnew311-inspect but they didn't detail this behavior.

Anyway I figured out we could actually just read the source file and offset the line number by 1 as long as we don't find a line starting with `def `. Do you think it's a good idea? I suppose it won't work for cythonized code but it's probably ok since this package is meant to document source files, and not files optimized for production?

ValentinFrancois commented on flask_selfdoc/autodoc.py:
```diff
@@ -1,19 +1,39 @@
+    Markup = markupsafe.Markup
 
 try:
-    from flask import _app_ctx_stack as stack
```
With python 3.11 and Flask 2.2.3 I was getting `_app_ctx_stack' is deprecated and will be removed in Flask 2.3.`
See https://github.com/pallets/flask/pull/4682

ValentinFrancois commented on flask_selfdoc/autodoc.py:
```diff
@@ -153,7 +254,7 @@ def generate(self, groups='all', sort=None):
                     rule="%s" % rule,
                     endpoint=rule.endpoint,
-                    docstring=func.__doc__,
+                    docstring=func.__doc__.strip(' ') if func.__doc__ else None,
```
I thought this could be more practical for teams where some developers have an IDE that automatically strips the trailing whitespaces

ValentinFrancois commented on flask_selfdoc/autodoc.py:
```diff
@@ -201,7 +302,10 @@ def html(self, groups='all', template=None, **context):
+    def json(self,
+             groups='all',
+             indent: Optional[int] = None,
+             separators: Optional[Tuple] = (',', ':')):
```
I'm adding some options that are passed to `json.dumps`, I found it practical to debug the JSON output comparisons in the unit tests. Default options produce the same output as `jsonify` (i.e. minified JSON)

ValentinFrancois commented on tests/config.py:
```diff
@@ -1,10 +1,3 @@
-# The old version chooses the location of the first decorator,
-# the new version chooses the location of the 'def' keyword.
-# We detect the version and support both.
-NEW_FN_OFFSETS = sys.version_info >= (3, 8)
```
we wouldn't need this anymore because I now try to add some offset if `stack()[1].lineno` doesn't point to a line starting by `def `

ValentinFrancois commented on tests/files/builtin.json:
```diff
@@ -1 +1,126 @@
-{"endpoints":[{"args":[],"docstring":"Return all posts.","methods":["GET","HEAD","OPTIONS"],"rule":"/"},{"args":[],"docstring":"Admin interface.","methods":["GET","HEAD","OPTIONS"],"rule":"/admin"},{"args":[["greeting","Hello"],["id",null]],"docstring":"Return the user for the given id.","methods":["GET","HEAD","OPTIONS"],"rule":"/greet/<greeting>/user/<int:id>"},{"args":[["id",-1]],"docstring":" ...
+{
```
this was to debug the unit tests, we can re-minify the output if you prefer

ValentinFrancois commented on tests/files/custom.json:
```diff
@@ -1,152 +1,152 @@
-    "rule": "/users"
-  }
-]
+    {
```
here I just replaced tabs by 2 spaces and removed the trailing whitespaces (more common format imo)

ValentinFrancois commented on tests/files/simple.json:
```diff
@@ -1 +1,185 @@
-[{"args":["None"],"defaults":{},"docstring":"Return all posts.","endpoint":"get_posts","location":{"filename":"%PATH%/examples/simple/blog.py","line":45},"methods":["GET","HEAD","OPTIONS"],"rule":"/"},{"args":["None"],"defaults":{},"docstring":"Admin interface.","endpoint":"admin","location":{"filename":"%PATH%/examples/simple/blog.py","line":113},"methods":["GET","HEAD","OPTIONS"],"rule":"/admin ...
+[
```
same as above, it was easier to debug like this but we can re-minify the output

ValentinFrancois commented on flask_selfdoc/autodoc.py:
```diff
@@ -22,6 +42,87 @@
+    return response
+
+
+def get_decorator_frame_info(frame) -> dict:
```
this is the fix for the line numbers that should work with any python version >= 3.6, with some documentation

ValentinFrancois commented on flask_selfdoc/autodoc.py:
```diff
@@ -44,7 +145,10 @@ def init_app(self, app):
 
     def teardown(self, exception):
-        ctx = stack.top  # noqa: F841
+        if _cv_app is not None:
```
I suppose this is some hack to avoid destroying the context or something like that?
I tried to do the same with the new app context object from Flask 2.3.

ValentinFrancois commented:
In the end the PR got bigger than I first planned, but all tests are green now! I updated the PR description to my latest changes

ValentinFrancois commented:
just made it work again with `pytest`, I suppose it didn't like the double inheritance trick

ValentinFrancois commented:
Sorry I didn't realize that 
`"${{ github.repositoryUrl }}@${{ github.head_ref }}"` would not work on your side,
 because `${{ github.repositoryUrl }}` points to the repository where the action is running, and not the repository where the PR comes from.

I switched to `"${{ github.event.pull_request.head.repo.git_url }}@${{ github.event.pull_request.head.ref }}"`,
 based on https://stackoverflow.com/a/65027014.

Now the `check_pip_install` tests should always install from the branch in the source repo (forked or not), which should work on your side too.

ValentinFrancois commented:
My bad, I forgot to do the same in the install step for Windows. Hoping it's the last iteration

ValentinFrancois commented:
Anything still blocking the merge?

I see one job that seems to be stuck for some reason
![image](https://user-images.githubusercontent.com/24223132/221426623-997c92d3-2905-44b0-a570-86edcb9cece7.png)

ValentinFrancois commented:
ping

jwg4 commented:
hi, sorry for delay, I've been very busy with some other work.

ValentinFrancois commented:
No worries, just let me know if there's anything else I can help with.
I don't really understand why a single job has stayed stuck this whole time: 
![image](https://user-images.githubusercontent.com/24223132/229106272-c71c3afa-3c77-4ce5-ae05-3ce5c42d2ad8.png)

jwaxy commented:
is there any progress in merging this?

ValentinFrancois committed:
fix Jinja imports

ValentinFrancois committed:
fix Flask compatibility

ValentinFrancois committed:
fix decorator line number

ValentinFrancois committed:
Add optional JSON doc formatting options and adapt unit tests

ValentinFrancois committed:
adapt CI actions

- increase range of python and Flask versions tested

- test correct install via pip

ValentinFrancois commented:
Oh thanks for the reminder! Just rebased into a few commits.

jwaxy commented:
And btw is this library compatible with flask-restful?

ValentinFrancois commented:
No, the auto documentation logic relies on the `Flask.route` decorator which is not used in flask-restful.

jwg4 reviewed (approved):
looks good

ValentinFrancois commented on .github/workflows/test.yml:
```diff
@@ -7,22 +7,104 @@ jobs:
-        run: poetry add flask==${{ matrix.flask-version }}
+        run: pip3 install poetry
+      - name: Overwrite Flask dependencies for legacy install
+        if: matrix.flask-version < '2.0' && matrix.flask-version != 'latest'
```
(Sorry, just seeing this now)

I wasn't sure how `matrix.flask-version < 2.0` would behave in the case where `matrix.flask-version == 'latest'`.
Now I think `matrix.flask-version < '2.0'` would be sufficient.

ValentinFrancois commented on .github/workflows/test.yml:
```diff
@@ -7,22 +7,104 @@ jobs:
+        os: [ubuntu-18.04, macos-latest]
+        exclude:
+          # starting from Flask 2.1.0, python 3.6 is no longer supported:
+          - flask-version: '2.1'
```
I guess now that python 3.6 is EOL we can indeed simplify the logic and stop testing on it

ValentinFrancois commented on .github/workflows/test.yml:
```diff
@@ -7,22 +7,104 @@ jobs:
+            exit 1
+          fi
+      - name: Try importing flask_selfdoc
+        run: python3 -c "import flask_selfdoc"
```
If I remember correctly the issue I fixed happened only at import time, but yes running an example is a good idea.
